shared/middleware/helpers/redis_table.rb

```diff
@@ -0,0 +1,108 @@
+# RedisTable provides a table abstraction that notifies clients of changes.
+# Operations include adding a row, updating and fetching a specific row
+# by id, and fetching all rows. Rows preserve the order of insertion.
+# # The implementation wraps a RedisProperty bag to store the table rows
+# # and a pub sub service to notify clients of changes.
+
+require_relative 'redis_property_bag'
+
+# @private
+class RedisTable
+  ROW_ID_KEY = "row_id"
+
+  def initialize(redis, pub_sub_api, shard_id, table_name)
+    @pub_sub_api = pub_sub_api
+
+    @shard_id = shard_id
+    @table_name = table_name
+    @props = RedisPropertyBag.new(redis, "#{shard_id}_#{table_name}")
+  end
+
+  # Inserts a new row, notifying other clients using the pubsub service.
+  #
+  # @param [Hash] value The hash for the new row.
+  # @return [Hash] The inserted value, including the new :id field.
+  def insert(value)
+    new_row_id = next_id
+    @props.set(new_row_id.to_s, value.to_json)
+    publish_change({:action => 'insert', :id => new_row_id})
+    merge_id(value, new_row_id)
+  end
+
+  # Converts the rows to an array ordered by ascending row id.
+  #
+  # @return [Array<String>]
+  def to_a
+    result = []
+    @props.to_hash.each do |key, value|
+      result << make_row(key, value) if key != ROW_ID_KEY
+    end
+    result.sort_by {|row| row['id']}
```

Comment:
Might be more rubyish to chain Enumerable operations.
```ruby
@props.to_hash.
  select { |key, value| key != ROW_ID_KEY }.
  collect { |key, value| make_row(key, value) }.
  sort_by { |row| row['id'] }
```
...though unless there's some cool optimization happening it might be less efficient.